File: backend/ts_template/devdonalds.ts
import express, { Request, Response } from "express";

// ==== Type Definitions, feel free to add or modify ==========================
interface cookbookEntry {
  name: string;
  type: string;
}

interface requiredItem {
  name: string;
  quantity: number;
}

interface recipe extends cookbookEntry {
  requiredItems: requiredItem[];
}

interface ingredient extends cookbookEntry {
  cookTime: number;
}

interface recipeSummary {
  name: string;
  cookTime: number;
  ingredients: requiredItem[];
}

// =============================================================================
// ==== HTTP Endpoint Stubs ====================================================
// =============================================================================
const app = express();
app.use(express.json());

// Store your recipes here!
const cookbook: Map<string, recipe | ingredient> = new Map();

// Task 1 helper (don't touch)
app.post("/parse", (req:Request, res:Response) => {
  const { input } = req.body;

  const parsed_string = parse_handwriting(input)
  if (parsed_string == null) {
    res.status(400).send("this string is cooked");
    return;
  }
  res.json({ msg: parsed_string });
  return;
});

// [TASK 1] ====================================================================
// Takes in a recipeName and returns it in a form that
const parse_handwriting = (recipeName: string): string | null => {
  if (!recipeName) return null;

  // Replace hyphens and underscores with spaces
  recipeName = recipeName.replace(/[-_]/g, " ");

  // Remove all non-letter characters except spaces
  recipeName = recipeName.replace(/[^a-zA-Z\s]/g, "");

  // Collapse multiple spaces into a single space
  recipeName = recipeName.replace(/\s+/g, " ");

  // Trim leading and trailing whitespace
  recipeName = recipeName.trim();

  // Make sure string is not empty after cleaning
  if (recipeName.length === 0) return null;

  // Capitalise first letter of each word, lowercase the rest
  recipeName = recipeName
    .split(" ")
    .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
    .join(" ");

  return recipeName;
}

// [TASK 2] ====================================================================
// Endpoint that adds a CookbookEntry to your magical cookbook
app.post("/entry", (req:Request, res:Response) => {
  const { input } = req.body;

  const success = addCookbookEntry(input);

  if (!success) {
    res.status(400).send();
    return;
  }
  res.status(200).send();
  return;
});

// Task 2 Backend function
const addCookbookEntry = (entry: any): boolean => {
  if (!entry || typeof entry.name !== "string" || typeof entry.type !== "string") {
    return false;
  }

  const { name, type } = entry;

  // Check valid type
  if (type !== "recipe" && type !== "ingredient") {
    return false;
  }

  // Check for unique name
  if (cookbook.has(name)) {
    return false;
  }

  // Testing ingredient now...
  if (type === "ingredient") {
    // checking cook time is not less than 0
    if (typeof entry.cookTime !== "number" || entry.cookTime < 0) {
      return false;
    }
    const newIngredient: ingredient = {
      name,
      type,
      cookTime: entry.cookTime
    }
    cookbook.set(name, newIngredient);
    return true;
  }

  // Testing recipe now...
  if (type === "recipe") {
    if (!Array.isArray(entry.requiredItems)) {
      return false;
    }

    const seen = new Set<string>();
    for (const item of entry.requiredItems) {
      if (typeof item.name !== "string" || typeof item.quantity !== "number") {
        return false;
      }
      // Checking required items are all unique
      if (seen.has(item.name)) {
        return false;
      }
      seen.add(item.name);
    }

    const newRecipe: recipe = {
      name,
      type,
      requiredItems: entry.requiredItems
    };
    cookbook.set(name, newRecipe);
    return true;
  }

  return false;
}

// [TASK 3] ====================================================================
// Endpoint that returns a summary of a recipe that corresponds to a query name
app.get("/summary", (req:Request, res:Request) => {
  // TODO: implement me
  res.status(500).send("not yet implemented!")

});

// Task 3 Backend function
const getRecipeSummary = (name: string): recipeSummary | null => {
  // Checking
  const entry = cookbook.get(name);
  if (!entry || entry.type === "ingredient") {
    return null;
  }

  const recipeFound = entry as recipe;
  const baseIngredients: requiredItem[] = [];
  let totalCookingTime = 0;

  for (const requiredItem of recipeFound.requiredItems) {
    // Check that recipe or ingredient exists
    const required = cookbook.get(requiredItem.name);
    if (!required) return null;

    if (required.type === "ingredient") {
      // Check whether ingredient already exists
      const existingIngredient = baseIngredients.find(ing => ing.name === requiredItem.name);
      if (!existingIngredient) {
        baseIngredients.push(requiredItem);
      } else {
        existingIngredient.quantity += requiredItem.quantity;
      }

      // Add to total cooking time
      const reqIngredient = required as ingredient;
      totalCookingTime += requiredItem.quantity * reqIngredient.cookTime;
    } else {

    }

  }
}

// =============================================================================
// ==== DO NOT TOUCH ===========================================================
// =============================================================================
const port = 8080;
app.listen(port, () => {
  console.log(`Running on: http://127.0.0.1:8080`);
});

File: backend/py_template/devdonalds.py
from dataclasses import dataclass
from typing import List, Dict, Union
from flask import Flask, request, jsonify
import re

# ==== Type Definitions, feel free to add or modify ===========================
@dataclass
class CookbookEntry:
	name: str

@dataclass
class RequiredItem():
	name: str
	quantity: int

@dataclass
class Recipe(CookbookEntry):
	required_items: List[RequiredItem]

@dataclass
class Ingredient(CookbookEntry):
	cook_time: int


# =============================================================================
# ==== HTTP Endpoint Stubs ====================================================
# =============================================================================
app = Flask(__name__)

# Store your recipes here!
cookbook = Dict[str, Union[Recipe, Ingredient]] = {}

# Task 1 helper (don't touch)
@app.route("/parse", methods=['POST'])
def parse():
	data = request.get_json()
	recipe_name = data.get('input', '')
	parsed_name = parse_handwriting(recipe_name)
	if parsed_name is None:
		return 'Invalid recipe name', 400
	return jsonify({'msg': parsed_name}), 200

# [TASK 1] ====================================================================
# Takes in a recipeName and returns it in a form that
def parse_handwriting(recipeName: str) -> Union[str | None]:
	if recipeName is None:
		return None

	# Replace hyphens with spaces
	recipeName = recipeName.replace("-", " ")

	# Replace underscores with spaces
	recipeName = recipeName.replace("_", " ")

	# Remove all non-letter characters expect spaces
	recipeName = re.sub(r"[^a-zA-Z\s]", "", recipeName)

	# Collapse multiple spaces into one
	recipeName = re.sub(r"\s+", " ", recipeName)

	# Trim leading and trailing spaces
	recipeName = recipeName.strip()

	# Return None if string is empty after cleaning
	if len(recipeName) == 0:
		return None

	# Capitalise each word
	words = recipeName.split(" ")
	words = [word.capitalize() for word in words]
	recipeName = " ".join(words)

	return recipeName


# [TASK 2] ====================================================================
# Endpoint that adds a CookbookEntry to your magical cookbook
@app.route('/entry', methods=['POST'])
def create_entry():
	data = request.get_json()

	success = add_cookbook_entry(data)

	if not success:
		return '', 400

	return '', 200


# Backend function
def add_cookbook_entry(data: dict) -> bool:
	global cookbook

	# Basic validation
	if not isinstance(data, dict):
		return False

	name = data.get("name")
	entry_type = data.get("type")

	if not isinstance(name, str) or not isinstance(entry_type, str):
		return False

	# Type must be recipe or ingredient
	if entry_type not in ["recipe", "ingredient"]:
		return False

	# Name must be unique
	if name in cookbook:
		return False


	# Testing ingredient...
	if entry_type == "ingredient":
		cook_time = data.get("cookTime")

		if not isinstance(cook_time, int) or cook_time < 0:
			return False
		cookbook[name] = Ingredient(name=name, cook_time=cook_time)
		return True


	# Testing recipe...
	if entry_type == "recipe":
		required_items = data.get("requiredItems")

		if not isinstance(required_items, list):
			return False

		seen_names = set()
		parsed_items: List[RequiredItem] = []

		for item in required_items:
			if not isinstance(item, dict):
				return False

			item_name = item.get("name")
			quantity = item.get("quantity")

			if not isinstance(item_name, str) or not isinstance(quantity, int):
				return False

			# Only one element per name
			if item_name in seen_names:
				return False

			seen_names.add(item_name)
			parsed_items.append(RequiredItem(name=item_name, quantity=quantity))

		cookbook[name] = Recipe(name=name, required_items=parsed_items)
		return True

	return False


# [TASK 3] ====================================================================
# Endpoint that returns a summary of a recipe that corresponds to a query name
@app.route('/summary', methods=['GET'])
def summary():
	# TODO: implement me
	return 'not implemented', 500


# =============================================================================
# ==== DO NOT TOUCH ===========================================================
# =============================================================================

if __name__ == '__main__':
	app.run(debug=True, port=8080)
